collector/metrics/metric_collector.go

```diff
@@ -0,0 +1,354 @@
+	"gopkg.in/tomb.v2"
+	"github.com/tidwall/gjson"
+	"github.com/prometheus/client_golang/prometheus"
+    "github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+const (
+	META_SERVER int = 0
+	REPLICA_SERVER int = 1
+)
+
+type Metric struct {
+	name string
+	value int64
+	mtype string
+	desc string
+	unit string
+}
+
+type Metrics []Metric
+
+type Attribute struct {
+	name string
+	value string
+}
+
+type Entity struct {
+	etype string
+	eid string
+	attributes string
+	metrics []Metric
+}
+
+var gaugeMetricsMap_ map[string]prometheus.GaugeVec
+var counterMetricsMap_ map[string]prometheus.CounterVec
+var histogramMetricsMap_ map[string]prometheus.Histogram
+var summaryMetricsMap_ map[string]prometheus.Summary
+
+var dataSource_ int
+
+type MetricCollector interface {
+	Start(tom *tomb.Tomb) error
+}
+
+func NewMetricCollector(dataSource int) MetricCollector {
+	dataSource_ = dataSource
+	gaugeMetricsMap_ = make(map[string]prometheus.GaugeVec, 128)
+	counterMetricsMap_ = make(map[string]prometheus.CounterVec, 128)
+	histogramMetricsMap_ = make(map[string]prometheus.Histogram, 128)
+	summaryMetricsMap_ = make(map[string]prometheus.Summary, 128)
+	initMetrics()
+	return &Collector{detectInterval: 10, detectTimeout: 10}
+}
+
+type Collector struct {
+	detectInterval  time.Duration
+	detectTimeout time.Duration
+}
+
+func (collector *Collector) Start(tom *tomb.Tomb) error {
+	ticker := time.NewTicker(collector.detectInterval)
+	for {
+		select {
+			case <-tom.Dying():
+				return nil
+			case <-ticker.C:
+				processAllMetaServerMetrics()
+				break
+			default:
+		}
+	}
+}
+
+func initMetrics() {
+	var addrs []string
+	if dataSource_ == META_SERVER {
+		addrs = viper.GetStringSlice("meta_servers")
+	} else {
+		addrs = viper.GetStringSlice("replica_servers")
+	}
+
+	for _, addr := range addrs {
+		data, err := getOneMetaServerMetrics(addr)
+		if err != nil {
+			log.Errorf(fmt.Sprintf("Get raw metrics from %s failed, err: %s", addr, err))
+			return
+		}
+		json_data := gjson.Parse(data)
+		for _, entity := range json_data.Array() {
+			for _, metric := range entity.Get("metrics").Array() {
+				var name string = metric.Get("name").String()
+				var mtype string = metric.Get("type").String()
+				var desc string = metric.Get("description").String()
+				log.Errorf("name:%s type:%s desc:%s", name, mtype, desc)
+				switch mtype {
+				case "counter":
+					if _, ok := counterMetricsMap_[name]; ok {
+						continue
+					}
+					counterMetric := promauto.NewCounterVec(prometheus.CounterOpts{
+						Name: name,
+						Help: desc,
+					}, []string{"level", "endpoint"})
+					counterMetricsMap_[name] = *counterMetric
+					break
+				case "gauge":
+					if _, ok := gaugeMetricsMap_[name]; ok {
+						continue
+					}
+					gaugeMetric := promauto.NewGaugeVec(prometheus.GaugeOpts{
+						Name: name,
+						Help: desc,
+					}, []string{"level", "endpoint"})
+					gaugeMetricsMap_[name] = *gaugeMetric
+					break
+				case "histogram":
```

Comment:
The new metrics framework dosen't has histogram type of metrics, so it can be removed.